---
title: "RnaSeq pipeline"
output:
  github_document:
    html_preview: true
    toc: true
    toc_depth: 4
---

## Getting Started

1. Create a new folder in fs3 or any other location with the project name - called the "project folder" in this README.   
2. Create a bin/ folder in the project folder.   
3. Download and copy the scripts to the bin/ folder.   

## Running the pipeline

### 1. Analysis info file
A central part of the pipeline is the **analysis info** file. It has information about the project location, the original run folder, the reference fasta, gtf and bed files, and the parameters used throughout the analysis.   

#### What is it?
The analysis info file is a simple .txt file, with each line providing information. Parameters are separated by the semicolon (i.e ";") character.   

The following is an example of the analysis info file:   

--------------------------------------------------------------------------------------------------   
**working_directory =** /mnt/cgs-fs3/Sequencing/Pipelines/RNASeq_pipeline/example_small_files/    
**run_folder =** /mnt/cgs-fs3/Sequencing/Pipelines/RNASeq_pipeline/example_small_files/xxx   
**run_samplesheet =** /mnt/cgs-fs3/Sequencing/Pipelines/RNASeq_pipeline/example_small_files/xxx    
**bcl2fastq_output =** /mnt/cgs-fs3/Sequencing/Pipelines/RNASeq_pipeline/example_small_files/fastq/   
**readType =** pairedEnd    
**reference_genome =**    
**trimgalore_params =** --gzip; --paired; --fastqc; --fastqc_args '--nogroup --extract'   
**ncores =** 8   
--------------------------------------------------------------------------------------------------   

The following is the explanation of the analysis info file:   

--------------------------------------------------------------------------------------------------   
**working_directory =** *\<path to directory of the analysis\>*   
**run_folder =** *\<path to the run folder\>*    
**run_samplesheet =** *\<sample sheet to be used to generate fastq files. This is created using the Illumina Expert Manager\>*    
**bcl2fastq_output =** *\<path to output of bcl2fastq. The defaults is fastq/ and the folder will be created automatically\>*    
**readType =** *\<either pairedEnd or singleEnd\>*     
**reference_genome =** *\<path to the STAR reference genome folder that will be used at the mapping step\>*   
**trimgalore_params =** *\<parameters to be passed to trim galore\>*    
**ncores =** *\<Number of cores to use to pararellize analysis\>*    
--------------------------------------------------------------------------------------------------    
   
<br>   

#### How to create one?

Use the bin/analysis_info.txt.   

```{r, engine = 'bash', eval = FALSE}
$ python bin/analysis_info.py
```

This will create a file analysis_info.txt, which you can open in a text editor and fill.

<br>   

### 2. Convert BCL to FASTQ files
Create fastq files (If you have more than one run, you can run this command for each run by changing the run_folder and run_samplesheet fields in the analysis_info.txt file)
```bash
$ python bin/run_bcl2fastq.py --analysis_info_file analysis_info.py
```
There is also a log file created called "bcl_log.txt" to check that the conversion didn't have any problems. 

#### Lexogen projects
For lexogen projects, we currently download the data from Basespace. This involves extra steps:   

1. Download fastq files from Basespace. Basespace splits the data in 4 files. An example of the file names is:   
    + sample1_S1_L001_R1_001.fastq   
    + sample1_S1_L002_R1_001.fastq   
    + sample1_S1_L003_R1_001.fastq   
    + sample1_S1_L004_R1_001.fastq    
If the run is paired end, there will also be files for R2:   
    + sample1_S1_L001_R2_001.fastq   
    + sample1_S1_L002_R2_001.fastq   
    + sample1_S1_L003_R2_001.fastq   
    + sample1_S1_L004_R2_001.fastq   
   
   
2. Concatenate the 4 files into one, so that at the end there is one file per sample (or two if its paired end). **The name of the concatenated file is important**. The only difference to the original file names, is that concatenated files do not contain the string *L00[1,2,3,4]*.   
    + sample1_S1_R1_001.fastq   
    + sample1_S1_R2_001.fastq   

3. It is possible that the **folder names** of the samples will not be in the right format. The scripts expect that each sample has a separate folder and that the folder names consist of the sample name and the _S[\\d] string. From the example above, the folder name would be:   
    + sample1_S1   
So the relative path be:  
    + sample1_S1/sample1_S1_R1_001.fastq
    + sample1_S1/sample1_S1_R2_001.fastq   
    
<br>

### 3. Create a sample names file
This should be a text file, with one sample name per line. The sample names consist of the sample name and the _S[\\d] string. From the example above, the sample name would be:   
- sample1_S1   
    
<br>

### 4. Organize working directory

### 5. Quality control of Fastq files
To run fastqc in all the samples, use the script bin/qcReads.py. 

Script: bin/qcReads.py

Arguments:   

--------------------------------------------------------------------------------------------------
-h, --help	show this help message and exit
-v, --version	show program's version number and exit
--analysis_info_file	ANALYSIS_INFO_FILE. Text file with details of the analysis. Default=analysis_info.txt
--in_dir	IN_DIR. Path to folder containing fastq files. Default=rawReads/
--out_dir	OUT_DIR. Path to out put folder. Default=rawReads/
--out_dir_report	OUT_DIR_REPORT. Path to out put folder. Default=Report/figure/data/
--sample_names_file	SAMPLE_NAMES_FILE. Text file with sample names to run. Default=sample_names.txt
--------------------------------------------------------------------------------------------------



