---
title: "RnaSeq pipeline"
output:
  github_document:
    html_preview: true
    toc: true
    toc_depth: 3
---

## Getting Started
Create a new folder in fs3 or any other location with the project name - called the "project folder" in this README.   
Create a bin/ folder in the project folder.   
Download and copy the scripts to the bin/ folder.   

## Running the pipeline

### Analysis info file
A central part of the pipeline is the **analysis info** file. It has information about the project location, the original run folder, the reference fasta, gtf and bed files, and the parameters used throughout the analysis.

The following is an example of the analysis info file:   


> **working_directory =** /mnt/cgs-fs3/Sequencing/Pipelines/RNASeq_pipeline/example_small_files/    
> run_folder = /mnt/cgs-fs3/Sequencing/Pipelines/RNASeq_pipeline/example_small_files/xxx   
> run_samplesheet = /mnt/cgs-fs3/Sequencing/Pipelines/RNASeq_pipeline/example_small_files/xxx    
> bcl2fastq_output = /mnt/cgs-fs3/Sequencing/Pipelines/RNASeq_pipeline/example_small_files/fastq/   
> readType = pairedEnd    
> reference_genome =    
> trimgalore_params = --gzip; --paired; --fastqc; --fastqc_args '--nogroup --extract'   
> ncores = 8   


The following is the explanation of the analysis info file:   

--------------------------------------------------------------------------------------------------
working_directory = *\<path to directory of the analysis\>*   
run_folder = *\<path to the run folder\>*    
run_samplesheet = *\<sample sheet to be used to generate fastq files. This is created using the Illumina Expert Manager\>*    
bcl2fastq_output = *\<path to the desired output of bcl2fastq. The defaults is fastq/ and the folder will be created automatically\>*    
readType = *\<either pairedEnd or singleEnd\>*     
reference_genome = *\<path to the STAR reference genome folder that will be used at the mapping step\>*   
trimgalore_params = *\<parameters to be passed to trim galore\>*    
ncores = *\<Number of cores to use to pararellize analysis\>*    
--------------------------------------------------------------------------------------------------

 
